### PR TITLE
Fix deployment hanging

### PR DESCRIPTION
**Pull request by sergioboi**

### CODE REVIEW

File: subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/runner/TestRunnerImplementation.kt [hunk 1]
```diff
@@ -21,8 +20,8 @@ class TestRunnerImplementation(
 
     private val logger = loggerFactory.create<TestRunner>()
 
-    override suspend fun runTests(tests: List<TestRunRequest>, scope: CoroutineScope): TestRunnerResult {
-        return withContext(scope.coroutineContext) {
+    override suspend fun runTests(tests: List<TestRunRequest>): TestRunnerResult {
+        return coroutineScope {
```
Comment (sergioboi) [inline subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/runner/TestRunnerImplementation.kt:24]: Suddenly, we have the opportunity to use `coroutineScope` in `suspend` functions
It reduces a little of boilerplate

File: subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt [hunk 2]
```diff
@@ -11,56 +11,62 @@ import com.avito.logger.create
...
+import kotlinx.coroutines.channels.consumeEach
 import kotlinx.coroutines.channels.distinctBy
 import kotlinx.coroutines.channels.filter
 import kotlinx.coroutines.channels.toList
 import kotlinx.coroutines.delay
 import kotlinx.coroutines.launch
 import kotlinx.coroutines.withContext
 
+@OptIn(ExperimentalCoroutinesApi::class)
 internal class KubernetesReservationClient(
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt:30]: Here we go! This file is an essential change of this PR

File: subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt [hunk 3]
```diff
@@ -11,56 +11,62 @@ import com.avito.logger.create
...
 
-    private val podsQueryIntervalMs = 5000L
-
-    override fun claim(
+    override suspend fun claim(
         reservations: Collection<ReservationData>,
         scope: CoroutineScope
     ): ReservationClient.ClaimResult {
-
+        require(reservations.isNotEmpty()) {
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt:48]: Add fail fast

File: subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt [hunk 4]
```diff
@@ -69,44 +75,46 @@ internal class KubernetesReservationClient(
...
+            @Suppress("DEPRECATION")
+            podsChannel
                 .filter { it.status.phase == POD_STATUS_RUNNING }
                 .distinctBy { it.metadata.name }
-
-            for (pod in uniqueRunningPods) {
-                launch {
-                    pod.bootDevice(serialsChannel)
+                .consumeEach { pod ->
+                    if (!serialsChannel.isClosedForSend) {
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/main/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClient.kt:109]: if `serialsChannel` is closed, for example from another part of a runner then we stop consuming pods

File: subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                // second
+                client.claim(listOf(ReservationData.stub()), this)
+            }
+        }
+        assertThat(exception.message)
+            .isEqualTo("Unable to start reservation job. Already started")
+    }
+
+    @Test
+    fun `get one pod then empty - success`() {
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt:85]: Emulates current behavior when we get no pods
We will close all channels and finish our job without exception
I want to rewrite error handling and kubernetes interaction in the next PR

File: subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        kubernetesApi.getPods = {
+            results.poll()
+        }
+        dispatcher.runBlockingTest {
+            client.claim(listOf(ReservationData.stub()), this)
+        }
+    }
+
+    @Test
+    fun `devices channel cancel - success`() {
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt:103]: success case when another part of the runner did her job and cancel devices channel

File: subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+            listOf(StubPod())
+        }
+        runBlocking {
+            val result = client.claim(listOf(ReservationData.stub()), this)
+            result.deviceCoordinates.cancel()
+        }
+    }
+
+    @Test
+    fun `claim then release - success`() {
```
Comment (sergioboi) [inline subprojects/gradle/runner/device-provider/src/test/kotlin/com/avito/android/runner/devices/internal/kubernetes/KubernetesReservationClientTest.kt:115]: happy path - checks that there are no hangs

File: subprojects/gradle/runner/device-provider/src/testFixtures/kotlin/com/avito/android/runner/devices/internal/kubernetes/FakeReservationDeploymentFactory.kt [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package com.avito.android.runner.devices.internal.kubernetes
+
+import com.avito.android.runner.devices.model.ReservationData
+import com.fkorotkov.kubernetes.apps.metadata
+import io.fabric8.kubernetes.api.model.apps.Deployment
+
+internal class FakeReservationDeploymentFactory : ReservationDeploymentFactory {
```
Comment (eugene-krivobokov) [inline subprojects/gradle/runner/device-provider/src/testFixtures/kotlin/com/avito/android/runner/devices/internal/kubernetes/FakeReservationDeploymentFactory.kt:7]: `Fake` -> `Stub` ?

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/gradle/runner/device-provider/src/testFixtures/kotlin/com/avito/android/runner/devices/internal/kubernetes/FakeReservationDeploymentFactory.kt (hunk 8 above)
Comment (sergioboi) [reply subprojects/gradle/runner/device-provider/src/testFixtures/kotlin/com/avito/android/runner/devices/internal/kubernetes/FakeReservationDeploymentFactory.kt:7]: I consider we will use it as a Fake in the future

**Pull request merged**